### PR TITLE
remove search globals

### PR DESCRIPTION
**Pull request by roccoblues**
**Goals**

* get rid of `OptSearchReverse` from `globals.c`
* don't use op codes (`OP_SEARCH_BACKWARDS`, ...) in `mutt_search_command()` and `mutt_search_alias_command()`
* remove `SearchPattern`, `LastSearch` and `LastSearchExpn` globals from `pattern/pattern.c`

**Idea**

I split the required search data into two parts:
1. State that needs to be kept across operations. This is what the globals `OptSearchReverse` and `SearchPattern`, `LastSearch`, `LastSearchExpn` currrently hold.
2. "one-time" search flags:
  a. `OP_SEARCH` and `OP_SEARCH_BACKWARDS` trigger the search functions to prompt the user for input.
  b. `OP_SEARCH_OPPOSITE` tells the search function to search in the opposite direction of what is currently the global state. 

For the state data I've added a new struct `SearchState` that holds the data previously in the globals. That struct itself is kept on `IndexSharedData`, `AliasMenuData` and `PostponeData`.

For the flags I added `SearchFlags` (`SEARCH_PROMPT`, `SEARCH_OPPOSITE`) that get computed in `op_search()` based on the operation.

-> Related discussion: https://github.com/neomutt/neomutt/discussions/3997

### CODE REVIEW

**Comment by flatcap** [comment]
That looks good!

> remove `SearchPattern`, `LastSearch` and `LastSearchExpn` globals from `pattern/pattern.c`

I was worried that this might prevent the user continuing a search in a different dialog.
I thought that you could:
- Search in Index `/john`
- Open Alias dialog
- Hit `<search-next>` to find `john`

But apparently that's not the case, so we needn't worry :-)

> The search state manipulation and flags computation clutter all `op_search()` functions.

I don't see that as a problem.
Each `op_search()` is self-contained and bridges the gap between GUI and library.

> Where/when to initialize the Search? On demand like here or once when IndexSharedData is created.

Hmm...
The Alias and Postponed dialogs are fairly short-lived.
The Index is very long-lived.

Given that the search object is near-trivial in size, I'd say create it once on startup of each Dialog.

> Extract search state/flags manipulation to separate method(s)?.

Is there _that_ much that could be extracted?

> Naming all the things ... struct Search, mutt_search_new(), mutt_search_free(), ...

I think I'd go for `struct SearchState`, `search_state_new()`, `_free()` and vars called `state` (in a new file).

**Comment by flatcap** [comment]
One larger problem for us to think about, is whether the Index and Alias search code should be in `libpattern`

Should `mutt_search_command()` live in `libindex` and rely on `libpattern`? (probably in a dedicated file)
Likewise `mutt_search_alias_command()` in `libalias`.

Or would it be possible to abstract the pattern-searchiness to work on arbitrary objects/data?
(this is probably a **lot** of work)

**Comment by roccoblues** [comment]
> Should `mutt_search_command()` live in `libindex` and rely on `libpattern`? (probably in a dedicated file)
> Likewise `mutt_search_alias_command()` in `libalias`.

Interesting. `mutt_search_command()`is also used in `libpostpone` are we fine with that?
On that note: does `SearchState` really belong in `libpattern`? Or should we move everything together into a `libsearch`?

File: pattern/pattern.c [hunk 1]
```diff
@@ -434,65 +430,58 @@ int mutt_pattern_func(struct MailboxView *mv, int op, char *prompt)
...
       buf_init(&err);
       OptSearchInvalid = true;
-      mutt_str_copy(LastSearch, buf_string(buf), sizeof(LastSearch));
-      mutt_str_copy(LastSearchExpn, buf_string(tmp), sizeof(LastSearchExpn));
+      buf_copy(state->string_expn, tmp);
       mutt_message(_("Compiling search pattern..."));
-      mutt_pattern_free(&SearchPattern);
+      mutt_pattern_free(&state->pattern);
       err.dsize = 256;
       err.data = mutt_mem_malloc(err.dsize);
```
Comment (roccoblues) [inline pattern/pattern.c:476]: Unrelated to my changes but do you know why we manually allocate the buffer data here? Shouldn't the `buf_*` functions simply allocate/grow the buffer if necessary?

Comment (flatcap) [reply pattern/pattern.c:476]: In the early days, `buf_new()` didn't exist, nor did the Buffer Pool.
In some code where the Buffer might not be needed (and to save an alloc) it was allocated on the stack.

With modern processors and lots of memory, we shouldn't be worrying about `malloc()`ing a few KB.
An interesting question is: Is the Buffer Pool really a saving?

**Review by flatcap — CHANGES_REQUESTED** [review]
Idea great.  Code great. Execution: **crash** :-)
`AliasMenuData.search_state` is `NULL` so `<search>` or `<search-next>` crashes.

**Comment by roccoblues** [comment]
> Execution: **crash** :-) `AliasMenuData.search_state` is `NULL` so `<search>` or `<search-next>` crashes.

Wow, thanks for checking! For me it works. 😕 Maybe we're triggering different code paths. Here's what I do:
* hit 'm' to compose a new mail
* hit 'tab' to open alias window
* hit '/' to trigger a search

-> this works

I find 4 different methods where `AliasMenuData` is initialized: `query_complete()`, `query_index()`, `alias_complete()` and `alias_dialog()`.

File: pattern/pattern.c (hunk 1 above)
Comment (roccoblues) [reply pattern/pattern.c:476]: > An interesting question is: Is the Buffer Pool really a saving?

For the `string`, `string_expn` in the `SearchState` or in general?

Comment (roccoblues) [reply pattern/pattern.c]: > With modern processors and lots of memory, we shouldn't be worrying about malloc()ing a few KB.

agreed, I've changed it

Comment (flatcap) [reply pattern/pattern.c]: > > Is the Buffer Pool really a saving?
> 
> For the `string`, `string_expn` in the `SearchState` or in general?

In general.
The _idea_ of a pool sounds good, but there aren't any _time-critical_ loops where we alloc/free Buffers.

> Shouldn't the `buf_*` functions simply allocate/grow the buffer if necessary?

Many of them don't: `if (!buf || !buf->data) FAIL;`
We just need a plan.

**Comment by flatcap** [comment]
> Here's what I do:
> 
>     * hit 'm' to compose a new mail
> 
>     * hit 'tab' to open alias window
> 
>     * hit '/' to trigger a search

Yeah, that's exactly what I was doing, I thought.
It still crashes.
I'll run it through the debugger, PC (post-cornflakes :-)

File: pattern/pattern.c (hunk 1 above)
Comment (roccoblues) [reply pattern/pattern.c]: > > > Is the Buffer Pool really a saving? 
> > 
> > For the `string`, `string_expn` in the `SearchState` or in general?
> 
> In general. The _idea_ of a pool sounds good, but there aren't any _time-critical_ loops where we alloc/free Buffers.

I can't really judge this. But I'm happy to rip it out if we decide to do it. ;-)

**Review by flatcap — APPROVED** [review]

**Pull request merged**